// fg-js/src/gaps/data.ts
"use strict";

import * as utils from '../utils';  
import {DataQuery, IRenderMeta, IDataQueryObj} from '../valueMgr';  
import {Gap, IGapData} from '../client/gapClassMgr';  
import {Component} from '../client/componentBase';  
import {IAstNode} from '../outerTypes';

export interface IDataParsedData extends IGapData {
	value: IDataQueryObj;
};

export default class GData extends Gap {

	type: string = "data";
	value: DataQuery;
	public static isVirtual = false; 

	constructor (parsedMeta: IDataParsedData, parent: Gap){
		super(parsedMeta, parent);
		this.value = new DataQuery(parsedMeta.value);
		this.paths = [this.value.path];
	};

	static parse(node: IAstNode, parents: IGapData[]): IGapData {
		if (node.tagName != "data"){
			return null;
		};		
		const value = utils.parsePath(node).resolve(parents).toJSON();
		const meta: IDataParsedData = {
			type: "data",
			value,
			eid: node.attrs.id || null
		};
		return meta;
	};

	render(context: Component, data: any, meta: IRenderMeta): string {
		const value = this.value.render(data, meta);
		return utils.renderTag({
			name: "span",
			attrs: {},
			innerHTML: value
		});
	};

	update(context: Component, meta: Gap, scopePath: any, value: any){
		const node = meta.getDom()[0];
		if (!node){
			
		};
		node.innerHTML = value;
		//highlight(node, [0xffffff, 0xffee88], 500);
	};

};
