// fg-js/src/gaps/data.ts
"use strict";

import * as utils from '../utils/utils';  
import {DataQuery, IRenderMeta, IDataQueryObj} from '../common/valueMgr';  
import {Gap, IGapData} from '../client/gapClassMgr';  
import {Component} from '../client/componentBase';  
import {IAstNode} from '../outerTypes';
import DataDispatcher from '../dataDispatcher';

export interface IDataParsedData extends IGapData {
};

export default class GData extends Gap {

	type: string = "data";
	public static isVirtual = false; 

	static parse(node: IAstNode, dataDispatcher: DataDispatcher, parents: IGapData[]): IGapData {
		if (node.tagName != "data"){
			return null;
		};		
		let dataSource = utils.parsePath(node);
		dataDispatcher.reg([dataSource]);		
		const meta: IDataParsedData = {
			line: node.line,
			type: "data",
			dataSource: dataSource.toJSON(),
			eid: node.attrs.id || null
		};
		return meta;
	};

	render(context: Component, data: any, meta: IRenderMeta): string {
		const value = this.dataSource.render(data.self, meta);
		return utils.renderTag({
			name: "span",
			attrs: {
				id: this.genId(context.id, meta)
			},
			innerHTML: value
		});
	};

	update(context: Component, meta: IRenderMeta, oldData: any, newData: any){
		const val = this.dataSource.update(oldData.self, newData.self, meta);
		if (val === null){
			return;
		};
		const dom = this.getDom(context.id, meta)[0];
		dom.innerHTML = val;
	};

};
