// fg-js/src/gaps/root.ts
"use strict";

import * as utils from '../utils';  
import {IRenderMeta} from '../valueMgr';  
import {Gap, IGapData} from '../client/gapClassMgr';  
import {Component} from '../client/componentBase';
import {IAstNode} from '../outerTypes';
import {Template, TplData} from '../tplMgr';
import DataDispatcher from '../dataDispatcher';

export interface IRootParsedData extends IGapData {
	content: TplData;
};

export default class GRoot extends Gap{
	fg: Component;
	content: Template;
	type: string = "root";

	public static isVirtual = true;

	constructor (parsedMeta: IRootParsedData){
		super(parsedMeta, null);
		this.content = new Template(parsedMeta.content, this); 
	};

	static parse(node: IAstNode, dataDispatcher: DataDispatcher): Gap{
		return null;
	};

	render(context: Component, data: any, meta: IRenderMeta): string{
		return this.content.render(context, data.content, meta);
	};

	update(context: Component, meta: IRenderMeta, oldData: any, newData: any){
		this.content.update(context, meta, oldData, newData);
	};

};
