// fg-js/src/gaps/scope.ts
"use strict";

import * as utils from '../utils';  
import * as valueMgr from '../valueMgr';  
import {Gap, IGapData} from '../client/gapClassMgr';  
import {Component} from '../client/componentBase';
import {IAstNode} from '../outerTypes'; 
import {Template, TplData} from '../tplMgr';
import * as anchorMgr from '../anchorMgr';
import {IDataPath, IDataQueryObj, IScope, IScopeData, DataQuery, IRenderMeta} from '../valueMgr';
import DataDispatcher from '../dataDispatcher';

export interface IScopeParsedData extends IGapData {
	scope: IScopeData;
	content: TplData;
};

export default class GScope extends Gap{
	content: Template;
	scope: IScope;
	items: Gap[];
	type: string = "scope";
	
	public static isVirtual = true;

	constructor (parsedMeta: IScopeParsedData, parent: Gap){
		super(parsedMeta, parent);	
		this.scope.value = this.dataSource;	
		this.content = new Template(parsedMeta.content, this);
	};

	static parse(node: IAstNode, dataDispatcher: DataDispatcher, parents: IGapData[], html?: string): IGapData {
		if (node.tagName !== "scope"){
			return null;
		};
		const scopeName: string = "scope";
		const dataSource = utils.parsePath(node);	
		const scopeDD = dataDispatcher.reg([dataSource], true, scopeName);	
		const meta: IScopeParsedData = {
			line: node.line,
			dataSource,
			type: "scope",
			eid: node.attrs.id || null,
			scope: {
				name: scopeName,
				value: dataSource.toJSON(),
			},
			content: null
		};
		meta.content = Template.read(node, scopeDD, null, parents.concat([meta]));
		return meta;
	};

	render(context: Component, data: any, meta: IRenderMeta): string {
		const parts: string[] = data.items.map((dataPart: any, id: number) => {			
			const partMeta: IRenderMeta = {
				dynamicPath: meta.dynamicPath.concat([id])
			};
			return this.content.render(context, dataPart, partMeta);
		});
		const anchorCode = anchorMgr.genCode(context, this);		
		return parts.join('\n') + anchorCode;
	};

	update(context: Component, meta: IRenderMeta, oldData: any, newData: any){
		const hash = context.objectStorage.hash.bind(context.objectStorage);
		if (hash(oldData) === hash(newData)){
			return;
		};
		newData.forEach((newItem: any, id: number) => {
			const oldItem = oldData[id];
			if (hash(oldItem) === hash(newItem)){
				return;
			};
			const rm: IRenderMeta = {
				dynamicPath: meta.dynamicPath.concat([id])
			}
			this.content.update(context, rm, oldItem, newItem);
		});
		
	};

};
