// fg-js/src/serverUtils.ts
"use strict";

import * as fs from 'fs';
import * as path from 'path';

export enum FileKind{file, dir};

export interface FileMapRes{
	path: string;
	kind: FileKind;
	data?: Buffer | string;
	stats?: fs.Stats;
	files?: string[];
};

export type FileMapFn = (input: FileMapRes)=>FileMapRes;

export function folderMapSync(src: string, dest: string, fn: FileMapFn){
	fs.readdirSync(src).forEach(filename => {
		const filePath = path.relative(src, filename);
		const stats = fs.statSync(filePath);
		if (stats.isFile()){
			const srcContent = fs.readFileSync(filePath);
			const res: FileMapRes = fn({
				kind: FileKind.file,
				path: filePath,
				data: srcContent,
				stats,
				files: null
			});
			fs.writeFileSync(res.path, res.data);
		};	
		if (stats.isDirectory()){
			const files = fs.readdirSync(filePath);
			const res: FileMapRes = fn({
				kind: FileKind.dir,
				path: filePath,
				data: null,
				stats,
				files
			});
			fs.mkdirSync(res.path);
		};		
	});
};

export function prefixLines(str: string, prefix: string, triggerFn?: Function): string{
	const lines: string[] = str.split('\n').map(function(line, id){
		if (!triggerFn || triggerFn(line, id, lines)){
			return prefix + line;
		};
		return line;
	});
	return lines.join('\n');
};

export function fileExist(path: string): boolean{
	try{
		fs.accessSync(path);
	}catch(e){
		return false;
	};
	return true;
};

export function forTree(treeObj: any, childProp: string, fn: Function){
	fn(treeObj);
	if (treeObj[childProp]){
		treeObj[childProp].forEach(function(node: any){
			forTree(node, childProp, fn);
		});
	};
};

export function getSubFolders(path: string): string[]{
	return fs.readdirSync(path).filter(function(subPath: string){
		const stat = fs.statSync(path + '/' + subPath);
		return stat.isDirectory();
	});
};

export function treeMap(treeObj: any, childProp: string, fn: Function): any{
    let res: any = {};
	res = fn(treeObj);
	if (treeObj[childProp]){
		treeObj[childProp].forEach(function(node: any, id: number){
			res[childProp][id] = treeMap(node, childProp, fn);
		});
	};
	return res;
};
